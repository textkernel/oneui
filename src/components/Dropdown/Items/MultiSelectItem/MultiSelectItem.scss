.MultiSelectItem {
    composes: DropdownItem from '../Item.scss';

    // TODO: format variants
}