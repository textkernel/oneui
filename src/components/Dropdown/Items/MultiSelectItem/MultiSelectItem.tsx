import React from 'react';
import {
    DropdownMenuCheckboxItem,
    DropdownMenuCheckboxItemProps,
} from '@radix-ui/react-dropdown-menu';
import { bem } from '../../../../utils';
import { VisualCheckbox } from '../../../Checkbox';
import { Text } from '../../../Text';
import styles from './MultiSelectItem.scss';

export interface Props extends DropdownMenuCheckboxItemProps {
    /** Id for the checkbox */
    id?: string;
    /** a variant to determine the look and feel of component */
    variant?: 'option' | 'select-all' | 'group-title';
    /** to add a priority badge before the label */
    hasPriority?: boolean;
    /** Checkbox status */
    isSelected?: boolean;
}

const { block } = bem('MultiSelectItem', styles);

export const MultiSelectItem = React.forwardRef<HTMLLIElement, Props>(
    (
        {
            children,
            isSelected = false,
            disabled = false,
            variant = 'option',
            hasPriority = false, // eslint-disable-line @typescript-eslint/no-unused-vars
            onCheckedChange,
            ...rest
        },
        ref
    ) => (
        <DropdownMenuCheckboxItem
            ref={ref}
            role="option"
            aria-selected={isSelected}
            tabIndex={disabled ? -1 : 0}
            checked={isSelected}
            onSelect={(e) => {
                e.preventDefault();
            }}
            onCheckedChange={onCheckedChange}
            {...rest}
            {...block({
                isSelected,
                disabled,
                variant,
                ...rest,
            })}
        >
            <VisualCheckbox checked={isSelected} />
            <Text inline>{children}</Text>
        </DropdownMenuCheckboxItem>
    )
);
