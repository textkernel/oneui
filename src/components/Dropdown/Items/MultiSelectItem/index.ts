export { MultiSelectItem, Props as MultiSelectItemProps } from './MultiSeletItem';
