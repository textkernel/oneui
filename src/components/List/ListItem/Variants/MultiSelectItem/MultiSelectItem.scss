.MultiSelectItem {
    composes: OneUI-label-text from global;
    padding: var(--space-150);

    &--variant_select-all {
        composes: OneUI-label-text-bold from global;
        border-bottom: 1px solid var(--color-border-subtle);

        &.MultiSelectItem--disabled {
            color: var(--color-text-disabled);
        }
    }

    &--variant_group-title {
        composes: OneUI-caption-upper-case-text-bold from global;
        padding-bottom: var(--space-50);
        color: var(--color-text-brand-default, #0079A7);

        &.MultiSelectItem--disabled {
            color: var(--color-text-brand-disabled);
        }
    }

    &:focus-visible {
        outline: 2px solid var(--color-info-20, #99CAFF);
    }

    &:has(label:hover),
    &--isHighlighted {
        background: var(--color-background-neutral-subtlest-hover, #E6E6E6);
    }

    &:has(label:active) {
        background: var(--color-background-neutral-subtlest-pressed, #CCC);
    }

    &--disabled {
        color: var(--color-text-disabled);

        &:focus-visible,
        &:has(label:hover),
        &.MultiSelectItem--isHighlighted,
        &:has(label:active) {
            background: var(--color-background-neutral-subtlest-disabled, #FFF);
        }
    }

    &--isSelected.MultiSelectItem--variant_group-title,
    &--isSelected.MultiSelectItem--variant_option {
        color: var(--color-text-selected-default, #004999);
    
        &.MultiSelectItem--disabled {
            color: var(--color-text-selected-disabled, #004999);
        }
    }

    &--variant_group-title,
    &--isSelected.MultiSelectItem--variant_option {
        background: var(--color-background-selected-subtlest-default, #E6F2FF);

        &:has(label:hover),
        &--isHighlighted {
            background: var(--color-background-selected-subtlest-hover, #CCE4FF);
        }
    
        &:has(label:active) {
            background: var(--color-background-selected-subtlest-pressed, #99CAFF);
        }
    
        &.MultiSelectItem--disabled {
            &:focus-visible,
            &:has(label:hover),
            &.MultiSelectItem--isHighlighted,
            &:has(label:active) {
                background: var(--color-background-selected-subtlest-default, #E6F2FF);
            }
        }
    }
}