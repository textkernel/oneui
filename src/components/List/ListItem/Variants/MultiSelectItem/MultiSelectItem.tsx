import React from 'react';
import { bem } from '../../../../../utils';
import { Props as ListItemProps } from '../../ListItem';
import { Checkbox } from '../../../../Checkbox';
import { Text } from '../../../../Text';
import { SPACE_KEY } from '../../../../../constants';
import styles from './MultiSelectItem.scss';

export interface Props
    extends Omit<ListItemProps, 'onClick' | 'onChange' | 'value' | 'highlightContext'> {
    /** A function to be called if the item is clicked */
    onChange?: (e: React.SyntheticEvent<HTMLInputElement | HTMLLIElement>, value) => void;
    /** Item identifier is used in {@link Dropdown} to select/navigate through children */
    value: unknown;
    /** Id for the checkbox */
    id?: string;
    /** a variant to determine the look and feel of component */
    variant?: 'option' | 'select-all' | 'group-title';
    /** to add a priority badge before the label */
    hasPriority?: boolean;
}

const { block } = bem('MultiSelectItem', styles);

export const MultiSelectItem = React.forwardRef<HTMLLIElement, Props>(
    (
        {
            children,
            isSelected = false,
            isHighlighted = false,
            onChange,
            id,
            value,
            disabled = false,
            passDisabledToLi = false,
            variant = 'option',
            hasPriority = false, // eslint-disable-line @typescript-eslint/no-unused-vars
            ...rest
        },
        ref
    ) => {
        const uId = React.useId();
        const idToUse = id || uId;

        const liProps: React.HTMLAttributes<HTMLLIElement> & { disabled?: boolean } = rest;
        if (passDisabledToLi) {
            liProps.disabled = disabled;
        }

        const handleChange = (e: React.SyntheticEvent<HTMLLIElement | HTMLInputElement>) => {
            e.stopPropagation();
            onChange?.(e, value);
        };

        const handleKeyDown = (e: React.KeyboardEvent<HTMLLIElement | HTMLInputElement>) => {
            e.stopPropagation();
            if ((e.target as HTMLElement).nodeName === 'LI' && e.key === SPACE_KEY) {
                handleChange(e);
            }
        };

        return (
            <li
                {...liProps}
                onKeyDown={handleKeyDown}
                ref={ref}
                role="option"
                aria-selected={isSelected}
                tabIndex={disabled ? -1 : 0}
                {...block({
                    isHighlighted,
                    isSelected,
                    disabled,
                    passDisabledToLi,
                    variant,
                    ...rest,
                })}
            >
                <Checkbox
                    id={idToUse}
                    checked={isSelected}
                    disabled={disabled}
                    onChange={handleChange}
                    tabIndex={-1}
                >
                    <>
                        {/* to be implemented: {hasPriority && <PriorityBadge />} */}
                        <Text inline>{children}</Text>
                    </>
                </Checkbox>
            </li>
        );
    }
);

MultiSelectItem.displayName = 'MultiSelectItem';
