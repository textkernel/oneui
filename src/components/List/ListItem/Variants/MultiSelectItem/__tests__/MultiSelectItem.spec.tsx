import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MultiSelectItem } from '../MultiSelectItem';

describe('MultiSelectItem', () => {
    it('should render correctly', () => {
        const view = render(
            <MultiSelectItem value="select_all" onChange={jest.fn()}>
                Select all
            </MultiSelectItem>
        );

        expect(view.container).toMatchSnapshot();
    });

    it('should call onChange callback correctly with keyboard navigation', async () => {
        const user = userEvent.setup();
        const handleChange = jest.fn();
        const value = 'select_all';
        render(
            <MultiSelectItem value={value} onChange={handleChange}>
                Select all
            </MultiSelectItem>
        );

        await user.keyboard('[Tab]');
        await user.keyboard(' ');

        expect(handleChange).toHaveBeenCalledWith(expect.any(Object), value);
    });

    it('should call onChange callback correctly with value when checkbox is clicked', async () => {
        const user = userEvent.setup();
        const handleChange = jest.fn();
        const value = 'select_all';
        render(
            <MultiSelectItem value={value} onChange={handleChange}>
                Select all
            </MultiSelectItem>
        );

        await user.click(screen.getByRole('checkbox'));

        expect(handleChange).toHaveBeenCalledTimes(1);
        expect(handleChange).toHaveBeenCalledWith(expect.any(Object), value);
    });

    it('should call onChange callback correctly with keyboard interaction on <li>', async () => {
        const user = userEvent.setup();
        const handleChange = jest.fn();
        const value = 'select_all';
        render(
            <MultiSelectItem value={value} onChange={handleChange}>
                Select all
            </MultiSelectItem>
        );

        await user.type(screen.getByRole('option'), ' ');

        expect(handleChange).toHaveBeenCalledTimes(1);
        expect(handleChange).toHaveBeenCalledWith(expect.any(Object), value);
    });

    it('should call onChange callback correctly with keyboard interaction on <input>', async () => {
        const user = userEvent.setup();
        const handleChange = jest.fn();
        const value = 'select_all';
        render(
            <MultiSelectItem value={value} onChange={handleChange}>
                Select all
            </MultiSelectItem>
        );

        await user.type(screen.getByRole('checkbox'), ' ');

        expect(handleChange).toHaveBeenCalledTimes(2);
        expect(handleChange).toHaveBeenCalledWith(expect.any(Object), value);
    });
});
