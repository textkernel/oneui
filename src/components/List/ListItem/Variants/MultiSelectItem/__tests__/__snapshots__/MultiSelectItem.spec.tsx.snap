// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectItem should render correctly 1`] = `
<div>
  <li
    aria-selected="false"
    class="MultiSelectItem"
    role="option"
    tabindex="0"
  >
    <div
      class="Checkbox"
    >
      <input
        class="Checkbox__input"
        id=":r0:"
        tabindex="-1"
        type="checkbox"
      />
      <label
        class="Checkbox__label"
        for=":r0:"
      >
        <span
          class="Checkbox__box"
        >
          <svg
            class="Checkbox__svg"
            height="8px"
            role="img"
            viewBox="0 0 12 9"
            width="12px"
          />
        </span>
        <span
          class="Text--size_medium Checkbox__text"
        >
          Select all
        </span>
      </label>
    </div>
  </li>
</div>
`;
